### PR TITLE
starting super socket

### PR DESCRIPTION
**Pull request by pdlla**
unit tests pass so it must be working right?

### CODE REVIEW

**Comment by pdlla** [comment]
just initial tcp listener server module that's mosty copy pasted code from learnyousomeerlang.
It still needs to delegate the connections to other workers based on handshake message. TODO.

File: src/p2p/supersocket.erl [hunk 1]
```diff
@@ -0,0 +1,34 @@
+%% this module contains a supervised socket server thingy
+-module(supersocket).
+-behaviour(supervisor).
```
Comment (antonbobkov) [inline src/p2p/supersocket.erl]: 😱

File: src/p2p/supersocket.erl [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+-export([start_socket_server/0]).
+-export([start_socket/0]).
+-export([init/1]).
+
+-include_lib("stdlib/include/assert.hrl").
+
+start_socket_server() ->
+  io:fwrite("starting socket server on port 5678\n"),
```
Comment (antonbobkov) [inline src/p2p/supersocket.erl]: why is the port# the same as my password ಠ_ಠ

File: src/p2p/supersocket.erl [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+init([]) ->
+  Port = 5678,
+  %% what is {packet, line} idk?
+  {ok, ListenSocket} = gen_tcp:listen(Port, [{active,once}, {packet,line}]),
+  spawn_link(fun empty_listeners/0),
+  {ok, {{simple_one_for_one, 60, 3600},[
+    {socket,
+    {supersocket_tcplistener, start_link, [ListenSocket]}, % pass the socket!
+    temporary, 1000, worker, [sockserv_serv]}
+  ]}}.
```
Comment (antonbobkov) [inline src/p2p/supersocket.erl]: is this lisp?

Comment (antonbobkov) [reply src/p2p/supersocket.erl]: (also i promise to have some serious comments soon)

**Comment by pdlla** [comment]
abort! this is an old commit D:

**Comment by pdlla** [comment]
ok, it's fine now.

File: src/p2p/supersocket.erl (hunk 2 above)
Comment (pdlla) [reply src/p2p/supersocket.erl]: for security :o

**Comment by pdlla** [comment]
oh noes, travis failing on socket tests 😱

File: src/p2p/supersocket.erl [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+start_socket() ->
+  supervisor:start_child(?MODULE, []).
+
+%% Start with 20 listeners so that many multiple connections can join at once
+%% without waiting on previous one to finish connecting.
+empty_listeners() ->
+  [start_socket() || _ <- lists:seq(1,20)],
```
Comment (antonbobkov) [inline src/p2p/supersocket.erl]: is this python?

**Comment by antonbobkov** [comment]
how do you feel about using udp instead of tcp? will (maybe) have simpler network code

File: src/p2p/supersocket_tcplistener.erl [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  ok = inet:setopts(AcceptSocket, [{active, once}]),
+
+  %% go to next step
+  {noreply, S#state{socket=AcceptSocket, next=handshake}}.
+
+
+%% Let's get rid of the white space and ignore whatever's after.
+%% makes it simpler to deal with telnet.
+line(Str) ->
+  hd(string:tokens(Str, "\r\n ")).
```
Comment (antonbobkov) [inline src/p2p/supersocket_tcplistener.erl]: \r ಠ_ಠ

**Comment by antonbobkov** [comment]
you just copied this code from someone, right?

**Comment by antonbobkov** [comment]
oh yeah, learnsomeerlang, right

**Comment by antonbobkov** [comment]
do you know why travis is failing? Does it not do sockets at all? Have you tried sockets on travis before?

File: src/p2p/supersocket.erl (hunk 4 above)
Comment (pdlla) [reply src/p2p/supersocket.erl]: copy pastaed code from learnyousomeerlang is what it is

Comment (pdlla) [reply src/p2p/supersocket.erl]: you mean haskell? D:

File: src/p2p/supersocket_tcplistener.erl (hunk 5 above)
Comment (pdlla) [reply src/p2p/supersocket_tcplistener.erl]: \ಠ_ಠ

**Comment by pdlla** [comment]
I haven' tried sockets on travis yet. I'll shuffle the ports around or something. idk. 
kinda surprised it doesn't work.

**Comment by pdlla** [comment]
I didn't copy all of it :O just like half of it :O

File: test/supersock_test.erl [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    {ok, Sock} = gen_tcp:connect("localhost", 8384, [binary, {active, false}]),
+    {ok, Msg} = gen_tcp:recv(Sock, 0),
+    %% ?debugFmt("recv ~p~n",[Msg]),
+    gen_tcp:send(Sock, "handshake_reply"),
+    gen_tcp:close(Sock)
+  end).
+
+supersock_test() ->
+  supersocket:start_socket_server(),
+  potato_haskell:for(fun() -> connect_test end, 1),
```
[the remark this answers is not in the document]
Comment (pdlla) [reply test/supersock_test.erl]: :scream::scream::scream::scream::scream::scream::scream::scream::scream::scream::scream::scream:

**Comment by antonbobkov** [comment]
wait. its passing the tests now????

**Comment by pdlla** [comment]
:O

**Comment by pdlla** [comment]
quick ship it!

**Comment by pdlla** [comment]
I'm bored D:

**Pull request merged**